firmware/defmt-semihosting/src/lib.rs

```diff
@@ -1,56 +1,142 @@
+        }
+    }
+
+    /// Acquire the defmt encoder.
+    fn acquire(&self) {
         // safety: Must be paired with corresponding call to release(), see below
         let restore = unsafe { critical_section::acquire() };
 
-        if TAKEN.load(Ordering::Relaxed) {
+        // NB: You can re-enter critical sections but we need to make sure
+        // no-one does that.
+        if self.taken.load(Ordering::Relaxed) {
             panic!("defmt logger taken reentrantly")
         }
 
-        // no need for CAS because interrupts are disabled
-        TAKEN.store(true, Ordering::Relaxed);
+        // no need for CAS because we are in a critical section
+        self.taken.store(true, Ordering::Relaxed);
+
+        // safety: accessing the cell is OK because we have acquired a critical
+        // section.
+        unsafe {
+            self.cs_restore.get().write(restore);
+            let encoder: &mut defmt::Encoder = &mut *self.encoder.get();
+            let handle: &mut Option<Stdout> = &mut *self.handle.get();
+            if handle.is_none() {
+                *handle = semihosting::io::stdout().ok();
+            }
+            encoder.start_frame(|b| {
+                if let Some(h) = handle {
+                    _ = h.write_all(b);
+                }
+            });
+        }
+    }
+
+    /// Release the defmt encoder.
+    unsafe fn release(&self) {
+        if !self.taken.load(Ordering::Relaxed) {
+            panic!("defmt release out of context")
+        }
+
+        // safety: accessing the cell is OK because we have acquired a critical
+        // section.
+        unsafe {
+            let encoder: &mut defmt::Encoder = &mut *self.encoder.get();
+            let handle: &mut Option<Stdout> = &mut *self.handle.get();
+            encoder.end_frame(|b| {
+                if let Some(h) = handle {
+                    _ = h.write_all(b);
+                }
+            });
+            let restore = self.cs_restore.get().read();
+            self.taken.store(false, Ordering::Relaxed);
+            // paired with exactly one acquire call
+            critical_section::release(restore);
+        }
+    }
+
+    /// Write bytes to the defmt encoder.
+    unsafe fn write(&self, bytes: &[u8]) {
+        if !self.taken.load(Ordering::Relaxed) {
+            panic!("defmt write out of context")
+        }
+
+        // safety: accessing the cell is OK because we have acquired a critical
+        // section.
+        unsafe {
+            let encoder: &mut defmt::Encoder = &mut *self.encoder.get();
+            let handle: &mut Option<Stdout> = &mut *self.handle.get();
+            encoder.write(bytes, |b| {
+                if let Some(h) = handle {
+                    _ = h.write_all(b);
+                }
+            });
+        }
+    }
+}
 
-        // safety: accessing the `static mut` is OK because we have acquired a critical section.
-        unsafe { CS_RESTORE = restore };
+unsafe impl Sync for SemihostingEncoder {}
```

Comment:
Why does `SemihostingEncoder` need to be marked as `Sync`?

Reply:
Because it's used in a `static` variable.